Add support for varchar(x) in aggregate functions

fiedukow opened the pull request:
+ small fixup with @Description in second commit

martint commented:
@cberner, can you review?

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SqlAggregationFunction.java:
```diff
@@ -91,39 +102,27 @@ public boolean isDeterministic()
     {
-        private final InternalAggregationFunction function;
-        private final String description;
+        private final BindableAggregationFunction aggregationSignature;
```
Why do we need the `BindableAggregationFunction` class? It seems like that functionality could be merged into this class

fiedukow replied:
As in later comment - BindableAggregation function is top level.
In the future, it requires refactor to only have one responsibility (of holding aggregation metadata).

cberner replied:
`AnnotationBased` doesn't seem like the right name. I'd name this `SqlAggregationFunctionAdapter` or something like that

fiedukow replied:
This class is removed. Done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -254,4 +230,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+        return methods.build();
+    }
+
+    public static class BindableAggregationFunction
```
I think this should be a top-level class

fiedukow replied:
Done.
I left is as part of AggregationCompiler before, as those classes are still strictly connected. I wanted to avoid another refactor here. I think in the future this class should not use static parer methods. Instead, the BindableAggregationFunction should take all elements it needs already parsed, instead of things like:
```
    private final Class<?> definitionClass;
    private final Class<?> stateClass;
    private final Method inputFunction;
    private final Method outputFunction;
```

I think it is something that may be done as followup, but I would like to integrate it as is for now, to avoid longer discussion right now.

cberner replied:
Looking at this more carefully, why do we need this class? It seems to only be used by `AnnotationBasedSqlAggregationFunction` which is now just a wrapper around `BindableAggregationFunction`. I didn't read through all the details, but it seems like it would be better to have `AnnotationBasedSqlAggregationFunction` call a method on `AggregationCompiler` to get a `AggregationMetadata` and then it could modify that structure and call the compiler again in `specialize()`

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:
```diff
@@ -126,22 +125,24 @@ public static void mergeRegressionState(RegressionState state, RegressionState o
+        return generateAggregationName(baseName, outputType.getTypeSignature(), inputTypes.stream().map(x -> x.getTypeSignature()).collect(Collectors.toList()));
+    }
+
+    public static String generateAggregationName(String baseName, TypeSignature outputType, List<TypeSignature> inputTypes)
```
Instead of deprecating the other form of this, can you move this change to a separate commit and just remove the other function?

fiedukow replied:
Done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateAverageAggregations.java:
```diff
@@ -79,15 +80,15 @@ public static void combine(ApproximateAverageState state, ApproximateAverageStat
     }
 
-    @OutputFunction(StandardTypes.VARCHAR)
+    @OutputFunction(VARCHAR_MAX_LENGTH)
```
Can you split this commit into one that does the refactoring to enable this, and then a second commit to change all the functions to `varchar(...)`?

cberner replied:
I don't think max_length is appropriate here. This outputs a varchar resulting from the `formatApproximateAverage` function, so you should be able to determine the max length of that

cberner commented:
Back to you for comments

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/BindableAggregationFunction.java:
```diff
@@ -0,0 +1,211 @@
+                .collect(toImmutableList());
+    }
+
+    private static List<AggregationMetadata.ParameterMetadata> getParameterMetadata(@Nullable Method method, List<Type> inputTypes)
```
static import `ParameterMetadata`

fiedukow replied:
Done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateAverageAggregations.java:
```diff
@@ -79,15 +79,15 @@ public static void combine(ApproximateAverageState state, ApproximateAverageStat
         else {
             String result = formatApproximateAverage(state.getSamples(), state.getMean(), state.getM2() / state.getCount(), confidence);
-            VARCHAR.writeString(out, result);
+            createUnboundedVarcharType().writeString(out, result);
```
This doesn't look right. The signature declares `varchar(57)`, but you're writing with the unbound type. I think you need to use `VarcharType.createVarcharType`

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateCountAggregation.java:
```diff
@@ -44,11 +43,11 @@ public static void combine(ApproximateCountState state, ApproximateCountState ot
     {
         Slice value = Slices.utf8Slice(formatApproximateResult(state.getCount(), countError(state.getSamples(), state.getCount()), confidence, true));
-        VARCHAR.writeSlice(out, value);
+        createUnboundedVarcharType().writeSlice(out, value);
```
same comment

fiedukow replied:
done

fiedukow commented:
Comments applied. Rebased to current master. Conflicts solved.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateCountColumnAggregations.java:
```diff
@@ -66,7 +66,7 @@ public static void combine(ApproximateCountState state, ApproximateCountState ot
+    @OutputFunction("varchar(57)")
     public static void output(ApproximateCountState state, double confidence, BlockBuilder out)
     {
         String result = formatApproximateResult(state.getCount(), countError(state.getSamples(), state.getCount()), confidence, true);
```
This needs to change the writing with `varchar(57)` also

fiedukow replied:
done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSumAggregations.java:
```diff
@@ -46,7 +46,7 @@ public static void combine(ApproximateDoubleSumState state, ApproximateDoubleSum
     }
 
-    @OutputFunction(StandardTypes.VARCHAR)
+    @OutputFunction("varchar(57)")
```
same comment

fiedukow replied:
done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSumAggregations.java:
```diff
@@ -78,7 +78,7 @@ public static void combine(ApproximateLongSumState state, ApproximateLongSumStat
     }
 
-    @OutputFunction(StandardTypes.VARCHAR)
+    @OutputFunction("varchar(57)")
```
same comment

fiedukow replied:
done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -48,98 +57,69 @@
-    public AggregationCompiler(TypeManager typeManager)
+    // This function should only be used for function matching for testing purposes.
+    // General purpose function matching is done through FunctionRegistry.
+    public static BindableAggregationFunction generateAggregationBindableFunction(Class<?> clazz)
```
Add `@VisibleForTesting`

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -48,98 +57,69 @@
-    public InternalAggregationFunction generateAggregationFunction(Class<?> clazz, Type returnType, List<Type> argumentTypes)
+    // This function should only be used for function matching for testing purposes.
+    // General purpose function matching is done through FunctionRegistry.
+    public static BindableAggregationFunction generateAggregationBindableFunction(Class<?> clazz, TypeSignature returnType, List<TypeSignature> argumentTypes)
```
Add `@VisibleForTesting`

fiedukow replied:
done.

cberner commented:
6110fe726e2743b101d6ae6ebdc5a745c87f2368 and 7aaf4408a48b24fee8d8bcf857e01517531af5a2 look good, if you want to reorder the commits I can merge those first.

Had some questions about the rest of the approach

fiedukow commented:
Commits reordered to be able to introduce 7aaf4408a48b24fee8d8bcf857e01517531af5a2 earlier.
I opened separate PR here https://github.com/prestodb/presto/pull/5515 for that one commit.
I'll remove this commit from this PR when it will land on master.

fiedukow commented:
6110fe726e2743b101d6ae6ebdc5a745c87f2368 requires logic changes in code that are introduced in other commits of this PR. This is not something that easily portable without other changes so I would like to keep it together with the rest here.

If you want I can separate description related changes from 4290c012bff8acc4bafcf8b6095732b5d0c66f84 but they will still require the logic shift from this commit.

fiedukow commented:
I removed AnnotationBasedSqlAggregationFunction class completely as you suggested before.
I misunderstood you before - this class was obviously redundant - thanks for the tip :-)

Back to you

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -254,4 +232,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+
+        public InternalAggregationFunction getOnlySpecialization()
+        {
+            return getOnlySpecialization(new TypeRegistry());
```
This won't work. Types from plugins will be missing if you call `new TypeRegistry()`

fiedukow replied:
I will add VisibleForTesting here as it should only be used in testing scenarios.
It is compatible with what we did there before.

cberner replied:
Ah, I see. Instead of just adding `VisibleForTesting` can you move this to a helper class in just the test code that needs it? It looks like most (all?) of the public statics that rely on this are only used in test/benchmarking code and so can be removed from the aggregation implementations. Perhaps it would be best to do that refactoring in a separate pre-commit to this one.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -254,4 +232,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+
+        public InternalAggregationFunction getOnlySpecialization(TypeManager typeManager)
+        {
+            // TODO check if default specialization exists
```
What does this TODO mean?

fiedukow replied:
There should be check that there is specialization that don't need any variables bound.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationMetadata.java:
```diff
@@ -257,51 +250,41 @@ public ParameterMetadata(ParameterType parameterType, Type sqlType)
             }
         }
 
+        public static ParameterMetadata forBlockIndexParameter()
```
All of these are unused

cberner replied:
I think you missed this one

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/type/VarcharType.java:
```diff
@@ -39,6 +40,11 @@ public static VarcharType createVarcharType(int length)
         return new VarcharType(length);
     }
 
+    public static TypeSignature getParametrizedVarcharSignature(String param)
```
typo, should be `getParameterizedVarcharSignature`. Also, I think the argument should be an `int` not a `String` to make the API safer

fiedukow replied:
It is ment to be used like:
getParameterizedVarcharSignature("x");

cberner replied:
ah yes, I see. I misread it

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationUtils.java:
```diff
@@ -126,14 +123,10 @@ public static void mergeRegressionState(RegressionState state, RegressionState o
     }
 
-    public static Type getOutputType(@Nullable Method outputFunction, AccumulatorStateSerializer<?> serializer, TypeManager typeManager)
+    public static TypeSignature getOutputTypeSignature(@Nullable Method outputFunction)
```
Since this is only used in one place and is so simple now, I would just inline it and remove the method

fiedukow replied:
done.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -48,98 +58,70 @@
+                                        new Signature(
+                                                name,
+                                                aggregationAnnotation.approximate() ? FunctionKind.APPROXIMATE_AGGREGATE : FunctionKind.AGGREGATE,
+                                                ImmutableList.of(), // TODO parse constrains from annotations
```
Lets implement this. Also, can you add a commit that ports something like `MaxAggregationFunction` to be annotation based to make sure this new `BindableAggregationFunction` can support everything that's needed for it?

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java:
```diff
@@ -254,4 +232,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+            Method combineFunction = getCombineFunction(definitionClass, stateClass);
+            AccumulatorStateFactory<?> stateFactory = new StateCompiler().generateStateFactory(stateClass, classLoader);
+
+            try {
```
Lets move all the unreflection into the constructor. Otherwise, if you get an exception it won't show up until someone submits a query that tries to use the aggregation function. You might need some new structure to hold the partial parameter metadatas, since the input types aren't known in the constructor

cberner commented:
Can you squash these three together? 4290c012bff8acc4bafcf8b6095732b5d0c66f84, 13850bb943be038b6091c706fa59e70a73335d41, 05d87306bc44082b776c7d061393d3f9b0b4a5a2

cberner commented:
Back to you for comments

fiedukow commented:
I missed some of your older comments, as they were hidden by github. I'll response for them as well, sorry for that.

fiedukow commented:
Ok. I started to answer all the big comments above but I missed the point there. I'll try to express this here as a summary.

I totally agree with you, that there are things to be improved here.
We can add support for all annotation (even though there are no use cases for them right now).
We can and should split everything into right classes and separate parsing, parsed data and InternalFunction nicely (note that AggregationMetadata does not do that, it's part of InternalAggregationFunction, not SqlAggregationFunction, and requires specific Types).

But my general thought here is that this patch does not break anything, adds what is required to complete VARCHAR(x) support and does not drastically increase mess with classes in this code.

I would love to those refactors but:
 1) This will mix refactor job with feature implementation
 2) This PR will become enormous and it is quite big already
 3) This will continue to block VARCHAR(x) complete support in presto

So my proposition is to focus on things that I may have broken here, to complete this PR as soon as possible.
Than I will try to refactor this as separate PR. Also it would be nice to have https://github.com/prestodb/presto/pull/5377 talked through before, as I think that many concept behind both refactors are going to be the same.

What do you think?
How can we limit scope of this PR?

cberner commented:
Ya, makes sense. I think leaving support for other annotations until later is fine. There is actually a use case though. The implementation of MAX/MIN is quite complicated and could be simplified with annotations.
I would like to see the usage of `getOnlySpecialization()` in the statics addressed though. That's not test code, so using something that's `@VisibleForTesting` isn't correct. Also, have them call a function that constructs a `new TypeRegistry()` isn't a good precedent because plugin authors may copy it and then be confused when it doesn't work due to a missing type from their plugin

fiedukow commented:
Rebased to master.
Removed commit that was already merged.
Fixed problem with getOnlySpecialization() by extracting InternalAggregationFunction creation to Utility class.

I think it's ready to be look at again, so back to you @cberner :-)

cberner commented:
Sorry, should have been more clear about what I meant should be done with the static fields. Nothing in the `src/main` directory should depend on something annotated with `@VisibleForTesting`. A method with that annotation should only be used from code in `src/test`, so I was suggesting that you remove those static fields completely and move them to a class in the `src/test` directory

fiedukow commented:
I understand you well but it is not that simple.
We use this method for creating InternalAggregationFunctions that are used for testing and benchmarking. For example here:
`com.facebook.presto.operator.aggregation.AverageAggregations.LONG_AVERAGE;`

If we want to move this to `src/test` we have to move those variables to `src/test` as well to avoid introducing dependency `src/main` -> `src/test`. If we will do that however we have to introduce dependency between `presto-benchmark` and `presto-main` (tests).  I thought `src/main/testing` is a place created exactly for such purposes. Alternatively we can copy all those variables and methods to both `presto-benchmark` and tests in `presto-main`. It may be a solution, but it is another refactor to make.

Just to be clear - here is how it was done before:
https://github.com/prestodb/presto/pull/5381/files#diff-aaed4e518e1ba3eedc34b06a90c6d493L55
If you think that was better in some way I can fallback to that, but I think it's exactly the same thing.

I really think it still falls into the category of "fixing the whole world" in the PR that has defined scope.

If you wan't to continue discussion please suggest concrete solution or lets try to catch up on Slack or hangout to reduce RTT (we have 9h time zone difference so every round of comments here takes 24h).

cberner commented:
Oh you're right, sorry I missed that we were already new'ing a TypeRegistry. I filed an issue to fix this instead: https://github.com/prestodb/presto/issues/5544

cberner commented:
Merged. Thanks!

fiedukow commented:
Thank you :-)
